From 6458239bc8f09c75548e7d6af491ef56eefad8f6 Mon Sep 17 00:00:00 2001
From: p4nc4k3z <163622641+p4nc4k3z@users.noreply.github.com>
Date: Sun, 18 Aug 2024 19:04:17 -0400
Subject: [PATCH 070/125] Support for multiple weighted field sprite variants
 (#75636)

The JSON required for terrain-aware sprites for fields in layering.json
previously demanded the use of a sprite array with weights, despite
our only supporting one sprite at a time and ignoring the weights. This
change makes the field syntax make sense, by introducing support for
supporting multiple field sprites, and respecting the weights by
introducing randomness seeded by map position.
---
 doc/TILESET.md     |  4 ++--
 src/cata_tiles.cpp | 21 +++++++++++----------
 2 files changed, 13 insertions(+), 12 deletions(-)

diff --git a/doc/TILESET.md b/doc/TILESET.md
index 9041de38dc..146a92de2b 100644
--- a/doc/TILESET.md
+++ b/doc/TILESET.md
@@ -514,7 +514,7 @@ This entry sets it so that the f_desk furniture if it contains either a pen or a
 
 `"layer": 100` this defines the order the sprites will draw in. 1 drawing first 100 drawing last (so 100 ends up on top). This only works for items, Fields are instead drawn in the order they are stacked on the tile.
 
-`"sprite": [{"id": "desk_pen_1", "weight": 2}, {"id": "desk_pen_2", "weight": 2}]` an array of the possible sprites that can display. For items multiple sprites can be provided with specific weights and will be selected at random.
+`"sprite": [{"id": "desk_pen_1", "weight": 2}, {"id": "desk_pen_2", "weight": 2}]` an array of the possible sprites that can display. Multiple sprites can be provided with specific weights and will be selected at random for each item.
 
 `"offset_x": 16`, `"offset_y": -48` optional sprite offset.
 
@@ -524,7 +524,7 @@ This entry sets it so that the f_desk furniture if it contains either a pen or a
 
 `"field": "fd_fire"` the field id. (only supported in field_variants)
 
-`"sprite": [{"id": "desk_fd_fire", "weight": 1}]` A field can have at most one sprite.
+`"sprite": [{"id": "desk_fd_fire", "weight": 1}]` an array of the possible sprites that can display. Multiple sprites can be provided with specific weights and will be selected at random based on map position.
 
 `"offset_x": 16`, `"offset_y": -48` optional sprite offset.
 
diff --git a/src/cata_tiles.cpp b/src/cata_tiles.cpp
index 95bce2cd5d..69b9706796 100644
--- a/src/cata_tiles.cpp
+++ b/src/cata_tiles.cpp
@@ -149,6 +149,11 @@ pixel_minimap_mode pixel_minimap_mode_from_string( const std::string &mode )
     return pixel_minimap_mode::solid;
 }
 
+auto simple_point_hash = []( const auto &p )
+{
+    return p.x + p.y * 65536;
+};
+
 } // namespace
 
 static int msgtype_to_tilecolor( const game_message_type type, const bool bOldMsg )
@@ -2773,10 +2778,6 @@ bool cata_tiles::draw_from_id_string_internal( const std::string &id, TILE_CATEG
         }
     }
 
-    auto simple_point_hash = []( const auto & p ) {
-        return p.x + p.y * 65536;
-    };
-
     // seed the PRNG to get a reproducible random int
     // TODO: faster solution here
     unsigned int seed = 0;
@@ -3649,8 +3650,8 @@ bool cata_tiles::draw_field_or_item( const tripoint &p, const lit_level ll, int
                         if( fld.id().str() == layer_var.id ) {
 
                             // get the sprite to draw
-                            // roll should be based on the maptile seed to keep visuals consistent
-                            int roll = 1;
+                            // roll is based on the maptile seed to keep visuals consistent
+                            int roll = simple_point_hash( p ) % layer_var.total_weight;
                             std::string sprite_to_draw;
                             for( const auto &sprite_list : layer_var.sprite ) {
                                 roll = roll - sprite_list.second;
@@ -3677,8 +3678,8 @@ bool cata_tiles::draw_field_or_item( const tripoint &p, const lit_level ll, int
                             if( fld.id().str() == layer_var.id ) {
 
                                 // get the sprite to draw
-                                // roll should be based on the maptile seed to keep visuals consistent
-                                int roll = 1;
+                                // roll is based on the maptile seed to keep visuals consistent
+                                int roll = simple_point_hash( p ) % layer_var.total_weight;
                                 std::string sprite_to_draw;
                                 for( const auto &sprite_list : layer_var.sprite ) {
                                     roll = roll - sprite_list.second;
@@ -3766,7 +3767,7 @@ bool cata_tiles::draw_field_or_item( const tripoint &p, const lit_level ll, int
                             const bool layer_nv = nv_goggles_activated;
 
                             // get the sprite to draw
-                            // roll should be based on the maptile seed to keep visuals consistent
+                            // roll is based on the item seed to keep visuals consistent
                             int roll = i.seed % layer_var.total_weight;
                             std::string sprite_to_draw;
                             for( const auto &sprite_list : layer_var.sprite ) {
@@ -3811,7 +3812,7 @@ bool cata_tiles::draw_field_or_item( const tripoint &p, const lit_level ll, int
                                 const bool layer_nv = nv_goggles_activated;
 
                                 // get the sprite to draw
-                                // roll should be based on the maptile seed to keep visuals consistent
+                                // roll is based on the item seed to keep visuals consistent
                                 int roll = i.seed % layer_var.total_weight;
                                 std::string sprite_to_draw;
                                 for( const auto &sprite_list : layer_var.sprite ) {
-- 
2.45.2

